pennylane/_qubit_device.py

```diff
@@ -29,100 +30,164 @@
+    do not need to be overwritten. Utility methods include:
+
+    * :meth:`~.expval`, :meth:`~.var`, :meth:`~.sample`: return expectation values,
+      variances, and samples of observables after the circuit has been rotated
+      into the observable eigenbasis.
+
     Args:
-        wires (int): number of subsystems in the quantum state represented by the device.
-            Default 1 if not specified.
-        shots (int): Number of circuit evaluations/random samples used to estimate
-            expectation values of observables. Defaults to 1000 if not specified.
+        wires (int): number of subsystems in the quantum state represented by the device
+        shots (int): number of circuit evaluations/random samples used to estimate
+            expectation values of observables
+        analytic (bool): If ``True``, the device calculates probability, expectation values,
+            and variances analytically. If ``False``, a finite number of samples set by
+            the argument ``shots`` are used to estimate these quantities.
     """
-    #pylint: disable=too-many-public-methods
+
+    # pylint: disable=too-many-public-methods
+    _asarray = staticmethod(np.asarray)
+
     def __init__(self, wires=1, shots=1000, analytic=True):
         super().__init__(wires=wires, shots=shots)
+
         self.analytic = analytic
+        """bool: If ``True``, the device supports exact calculation of expectation
+        values, variance, and probabilities. If ``False``, samples are used
+        to estimate the statistical quantities above."""
+
+        self._wires_used = set()
+        """set[int]: wires acted on by quantum operations and observables"""
+
+        self._memory = False
+        """bool: if True, indicates to the device that samples should be returned.
+        This is required if samples are being returned by the QNode, or if the device is in
+        non-analytic mode."""
 
-        self._state = None
-        self._rotated_prob = None
-        self._wires_used = None
-        self._memory = None
         self._samples = None
+        """None or array[int]: If :attr:`~._memory` is True, stores the samples
+        generated by the device *after* rotation to diagonalize the observables."""
 
+        self.reset()
 
     def reset(self):
         """Reset the backend state.
 
         After the reset, the backend should be as if it was just constructed.
         Most importantly the quantum state is reset to its initial value.
         """
-        self._state = None
-        self._rotated_prob = None
-        self._wires_used = None
-        self._memory = None
+        self._wires_used = set()
+        self._memory = False
         self._samples = None
 
-    def execute(self, queue, observables, parameters=None):
-        """Execute a queue of quantum operations on the device and then measure the given observables.
+    def execute(self, circuit):
```

Comment:
Maybe it would be better to have the object called ``circuit_graph`` in ``execute`` such that no ambiguity arises while reading the code